webapi/src/main/scala/org/knora/webapi/responders/v2/ResourcesResponderV2.scala

```diff
@@ -2246,4 +2254,313 @@ class ResourcesResponderV2(responderData: ResponderData) extends ResponderWithSt
+      sortedProjectHistory: Seq[ResourceAndValueHistoryV2] = projectHistory.flatten.sortBy(_.versionDate)
+
+    } yield ResourceAndValueVersionHistoryResponseV2(projectHistory = sortedProjectHistory)
+
+  /**
+    * Returns the full history of a resource as events.
+    *
+    * @param resourceFullHistRequest the version history of a resource.
+    * @return the full history of resource as sequence of [[ResourceAndValueHistoryV2]].
+    */
+  def getResourceHistoryEvents(
+      resourceFullHistRequest: ResourceFullHistoryGetRequestV2): Future[Seq[ResourceAndValueHistoryV2]] = {
+
+    val resourceHist = resourceFullHistRequest.resourceVersionHistory.reverse
+    // Collect the full representations of the resource for each version date
+    val histories: Seq[Future[(ResourceHistoryEntry, ReadResourceV2)]] = resourceHist.map { hist =>
+      for {
+        fullRepresentations <- getResourceAtGivenTime(
+          resourceIri = resourceFullHistRequest.resourceIri,
+          versionHist = hist,
+          featureFactoryConfig = resourceFullHistRequest.featureFactoryConfig,
+          requestingUser = resourceFullHistRequest.requestingUser
+        )
+      } yield fullRepresentations
+    }
+    for {
+      fullReps: Seq[(ResourceHistoryEntry, ReadResourceV2)] <- Future.sequence(histories)
+
+      // Create an event for the resource at creation time
+      (creationTimeHist, resourceAtCreation) = fullReps.head
+      resourceCreateEvent: ResourceAndValueHistoryV2 = getResourceAtCreationDate(resourceAtCreation, creationTimeHist)
+      histEvents: Seq[ResourceAndValueHistoryV2] = Seq(resourceCreateEvent)
+
+      // Add history events of values
+      resourceAtValueChanges = fullReps.tail
+
+      // For each value version, form an event
+      valuesEvents: Seq[ResourceAndValueHistoryV2] = resourceAtValueChanges.flatMap {
+        case (versionHist, readResource) => getValueAtGivenVersionDate(readResource, versionHist, fullReps)
+      }
+
+    } yield histEvents ++ valuesEvents
+  }
+
+  /**
+    * Returns the full representation of a resource at a given date.
+    *
+    * @param resourceIri                the IRI of the resource.
+    * @param versionHist                the history info of the version; i.e. versionDate and author.
+    * @param featureFactoryConfig       the feature factory configuration.
+    * @param requestingUser             the user making the request.
+    * @return the full representation of the resource at the given version date.
+    */
+  private def getResourceAtGivenTime(resourceIri: IRI,
+                                     versionHist: ResourceHistoryEntry,
+                                     featureFactoryConfig: FeatureFactoryConfig,
+                                     requestingUser: UserADM): Future[(ResourceHistoryEntry, ReadResourceV2)] =
+    for {
+      resourceFullRepAtCreationTime: ReadResourcesSequenceV2 <- (responderManager ? ResourcesGetRequestV2(
```

Comment:
Instead of the responder sending itself a message, you could just call `getResourcesV2` directly.

Reply:
yes, I am not consistent in that, sometimes I send a message, sometimes I just call the function. It was done for sake of the debugger tool, I will refactor this.